File: src/functions/convertPitchClass.ts
import gcd from "./gcd";

/**
 * Converts a decimal number to its closest fractional representation.
 * 
 * This function finds the best fractional approximation of a decimal number
 * by testing denominators up to 99 and selecting the fraction with minimal error.
 * This is crucial for representing pitch ratios in just intonation systems.
 * 
 * @param decimal - The decimal number to convert to a fraction
 * @returns String representation of the simplified fraction
 * 
 * @example
 * decimalToFraction(1.5) // Returns "3/2"
 * decimalToFraction(1.25) // Returns "5/4"
 */
function decimalToFraction(decimal: number) {
  let bestNumerator = 1;
  let bestDenominator = 1;
  let minError = Math.abs(decimal - bestNumerator / bestDenominator);

  // Test denominators from 1 to 99 to find the best approximation
  for (let d = 1; d <= 99; d++) {
    const n = Math.round(decimal * d);
    const error = Math.abs(decimal - n / d);
    
    // If this fraction is more accurate, use it
    if (error < minError) {
      minError = error;
      bestNumerator = n;
      bestDenominator = d;
    }
  }

  // Simplify the fraction using GCD
  const gcdValue = gcd(bestNumerator, bestDenominator);
  return `${bestNumerator / gcdValue}/${bestDenominator / gcdValue}`;
}

/**
 * Converts a frequency value to its equivalent MIDI note number.
 * 
 * Uses the standard MIDI tuning where A4 = 440 Hz = MIDI note 69.
 * This conversion is essential for interfacing with MIDI systems and
 * calculating pitch relationships.
 * 
 * @param frequency - The frequency in Hz
 * @returns The equivalent MIDI note number (can be fractional)
 * 
 * @example
 * frequencyToMidiNoteNumber(440) // Returns 69 (A4)
 * frequencyToMidiNoteNumber(523.25) // Returns ~72 (C5)
 */
export function frequencyToMidiNoteNumber(frequency: number): number {
  // MIDI note number formula: 69 + 12 * log2(frequency / 440)
  const midi = 69 + 12 * Math.log2(frequency / 440);
  return midi;
}

/**
 * Converts pitch class values between different representation formats.
 * 
 * This is the core conversion function that enables the system to work with
 * different pitch notations (fractions, cents, decimal ratios, string lengths).
 * It converts any input format to all other formats, enabling seamless
 * interoperability between different musical tuning systems.
 * 
 * @param originalValue - The input value as a string
 * @param inputType - The format of the input value
 * @param stringLength - Reference string length for string length calculations
 * @param referenceFrequency - Reference frequency for frequency calculations
 * @returns Object with all converted formats, or null if conversion fails
 * 
 * @example
 * // Convert a fraction to all formats
 * convertPitchClass("3/2", "fraction", 100, 440)
 * // Returns: { fraction: "3/2", decimal: "1.5", cents: "701.955", ... }
 * 
 * @example
 * // Convert cents to all formats  
 * convertPitchClass("200", "cents", 100, 440)
 * // Returns all equivalent representations of 200 cents
 */
export default function convertPitchClass(originalValue: string, inputType: "fraction" | "cents" | "decimalRatio" | "stringLength", stringLength: number, referenceFrequency: number) {
  // Initialize variables for all representation formats
  let fractionVal = "";
  let decimalVal = 0;
  let centsVal = 0;
  let stringLenVal = 0;
  let freqVal = 0;

  try {
    if (inputType === "fraction") {
      // Parse fraction format (e.g., "3/2")
      const [num, denom] = originalValue.split("/").map(Number);
      if (isNaN(num) || isNaN(denom) || denom === 0) return null;
      
      fractionVal = originalValue;
      const dec = num / denom;
      decimalVal = dec;
      
      // Convert to other formats using mathematical relationships
      centsVal = 1200 * Math.log2(dec);           // Cents = 1200 * log2(ratio)
      stringLenVal = stringLength / dec;          // String length inversely proportional to frequency
      freqVal = referenceFrequency * dec;         // Frequency proportional to ratio
      
    } else if (inputType === "cents") {
      // Parse cents format (e.g., "701.955")
      const c = parseFloat(originalValue);
      centsVal = c;
      
      // Convert cents to ratio, then to other formats
      decimalVal = Math.pow(2, c / 1200);         // Ratio = 2^(cents/1200)
      fractionVal = decimalToFraction(decimalVal);
      stringLenVal = stringLength / decimalVal;
      freqVal = referenceFrequency * decimalVal;
      
    } else if (inputType === "decimalRatio") {
      // Parse decimal ratio format (e.g., "1.5")
      const dec = parseFloat(originalValue);
      decimalVal = dec;
      
      // Convert decimal to other formats
      fractionVal = decimalToFraction(dec);
      centsVal = 1200 * Math.log2(dec);
      stringLenVal = stringLength / dec;
      freqVal = referenceFrequency * dec;
      
    } else if (inputType === "stringLength") {
      // Parse string length format (e.g., "66.67")
      const slVal = parseFloat(originalValue);
      stringLenVal = slVal;
      
      // Calculate frequency ratio from string length ratio
      const ratio = stringLength / slVal;  // Inverse relationship
      decimalVal = ratio;
      
      if (ratio <= 0 || !isFinite(ratio)) return null;
      
      // Convert ratio to other formats
      fractionVal = decimalToFraction(ratio);
      centsVal = 1200 * Math.log2(ratio);
      freqVal = referenceFrequency * ratio;
    }
    
    // Return object with all converted formats
    return {
      fraction: fractionVal,
      decimal: decimalVal.toString(),
      cents: centsVal.toString(),
      stringLength: stringLenVal.toString(),
      frequency: freqVal.toString(),
    };
  } catch {
    // Return null if any conversion fails
    return null;
  }
}

/**
 * Shifts a pitch class base value to a different octave.
 * 
 * This function transposes a pitch class value by octaves, handling the conversion
 * appropriately for different input formats. It's used when generating pitch classes
 * across multiple octaves from a single octave template.
 * 
 * @param baseValue - The original pitch class value as a string
 * @param inputType - The format of the input value
 * @param targetOctave - The target octave (0, 1, 2, 3, or 4)
 * @returns The shifted value in the same format as the input
 * 
 * @example
 * // Shift a fraction up one octave
 * shiftPitchClassBaseValue("3/2", "fraction", 2) // Returns "3/1" (doubled)
 * 
 * @example
 * // Shift cents up one octave
 * shiftPitchClassBaseValue("701.955", "cents", 2) // Returns "1901.955" (+1200 cents)
 */
export function shiftPitchClassBaseValue(baseValue: string, inputType: "fraction" | "decimalRatio" | "cents" | "stringLength", targetOctave: 0 | 1 | 2 | 3 | 4): string {
  // Octave 1 is the reference - no shift needed
  if (targetOctave === 1) return baseValue;

  // Calculate octave steps from the reference octave 1
  // octave 0 => -1 step (-12 semitones)
  // octave 1 => 0 steps (reference)  
  // octave 2 => +1 step (+12 semitones)
  // octave 3 => +2 steps (+24 semitones)
  // octave 4 => +3 steps (+36 semitones)
  const octaveSteps = targetOctave - 1;

  try {
    if (inputType === "fraction") {
      // For fractions: multiply by 2^octaveSteps
      // Going up an octave doubles the frequency ratio
      const [num, den] = baseValue.split(/[:/]/).map(Number);
      if (!den || isNaN(num) || isNaN(den)) return baseValue;

      let newNum = num * Math.pow(2, octaveSteps);
      let newDen = den;

      // Handle negative octave steps (going down) by adjusting denominator instead
      if (octaveSteps < 0) {
        const divisor = Math.pow(2, -octaveSteps);
        newDen = den * divisor;
        newNum = num;
      }

      // Simplify the resulting fraction
      const g = gcd(newNum, newDen);
      return `${newNum / g}/${newDen / g}`;
      
    } else if (inputType === "decimalRatio") {
      // For decimal ratios: multiply by 2^octaveSteps
      const dec = parseFloat(baseValue);
      const shifted = dec * Math.pow(2, octaveSteps);
      return shifted.toString();
      
    } else if (inputType === "cents") {
      // For cents: add 1200 * octaveSteps
      // Each octave = 1200 cents in equal temperament
      const c = parseFloat(baseValue);
      const shifted = c + 1200 * octaveSteps;
      return shifted.toFixed(3);
      
    } else if (inputType === "stringLength") {
      // For string length: divide by 2^octaveSteps
      // Higher frequency (higher octave) = shorter string length
      // Inverse relationship: frequency * 2 = string length / 2
      const slVal = parseFloat(baseValue);
      const multiplier = Math.pow(2, -octaveSteps);
      const shifted = slVal * multiplier;
      return shifted.toString();
    }
  } catch (err) {
    console.error("Error shifting pitch class:", err);
    return baseValue; // Return original value on error
  }

  return baseValue;
}

File: python/functions/convert_pitch_class.py
"""
Conversion utilities for pitch class data between different formats.

This module provides functions to convert pitch class data between various representations:
- Fractions (e.g., "3/2")
- Cents (e.g., 702.0)
- Decimal ratios (e.g., 1.5)
- MIDI note numbers (e.g., 60)
"""

import math
from typing import Union
from .detect_pitch_class_type import detect_pitch_class_type


def convert_pitch_class(value: Union[str, int, float], target_format: str) -> Union[str, float, int]:
    """
    Converts a pitch class value from its current format to the target format.
    
    This function automatically detects the input format and converts to the requested
    target format. It handles conversions between fractions, cents, decimal ratios,
    and MIDI note numbers.
    
    Args:
        value: The pitch class value to convert
        target_format: Target format ("fraction", "cents", "decimal", "midi")
        
    Returns:
        The converted value in the target format
        
    Examples:
        >>> convert_pitch_class("3/2", "cents")
        701.955
        >>> convert_pitch_class(702.0, "fraction")
        "3/2"
        >>> convert_pitch_class(1.5, "midi")
        67  # Assuming C4 as reference
    """
    # First detect the current format
    current_format = detect_pitch_class_type(value)
    
    # If already in target format, return as-is (with type conversion if needed)
    if current_format == target_format:
        if target_format == "fraction":
            return str(value)
        elif target_format == "midi":
            return int(float(value))
        else:
            return float(value)
    
    # Convert to decimal ratio as intermediate format
    decimal_ratio = _to_decimal_ratio(value, current_format)
    
    # Convert from decimal ratio to target format
    return _from_decimal_ratio(decimal_ratio, target_format)


def _to_decimal_ratio(value: Union[str, int, float], source_format: str) -> float:
    """
    Converts a value from any format to decimal ratio.
    
    Args:
        value: The value to convert
        source_format: The source format of the value
        
    Returns:
        The value as a decimal ratio
    """
    if source_format == "decimal":
        return float(value)
    
    elif source_format == "fraction":
        numerator, denominator = map(int, str(value).split("/"))
        return numerator / denominator
    
    elif source_format == "cents":
        # Convert cents to ratio: ratio = 2^(cents/1200)
        return pow(2, float(value) / 1200.0)
    
    elif source_format == "midi":
        # Convert MIDI to ratio relative to MIDI note 60 (middle C)
        # Each semitone is a ratio of 2^(1/12)
        midi_value = int(value)
        semitones_from_c4 = midi_value - 60
        return pow(2, semitones_from_c4 / 12.0)
    
    else:
        raise ValueError(f"Unknown source format: {source_format}")


def _from_decimal_ratio(decimal_ratio: float, target_format: str) -> Union[str, float, int]:
    """
    Converts a decimal ratio to the target format.
    
    Args:
        decimal_ratio: The decimal ratio to convert
        target_format: The target format
        
    Returns:
        The value in the target format
    """
    if target_format == "decimal":
        return decimal_ratio
    
    elif target_format == "fraction":
        return _decimal_to_fraction(decimal_ratio)
    
    elif target_format == "cents":
        # Convert ratio to cents: cents = 1200 * log2(ratio)
        if decimal_ratio <= 0:
            raise ValueError("Cannot convert non-positive ratio to cents")
        return 1200.0 * math.log2(decimal_ratio)
    
    elif target_format == "midi":
        # Convert ratio to MIDI note relative to middle C (60)
        if decimal_ratio <= 0:
            raise ValueError("Cannot convert non-positive ratio to MIDI")
        semitones_from_c4 = 12.0 * math.log2(decimal_ratio)
        return int(round(60 + semitones_from_c4))
    
    else:
        raise ValueError(f"Unknown target format: {target_format}")


def _decimal_to_fraction(decimal: float, max_denominator: int = 99) -> str:
    """
    Converts a decimal ratio to its closest fractional representation.
    
    Args:
        decimal: The decimal number to convert
        max_denominator: Maximum denominator to test
        
    Returns:
        String representation of the fraction
    """
    from .gcd import gcd
    
    best_numerator = 1
    best_denominator = 1
    min_error = abs(decimal - best_numerator / best_denominator)

    for d in range(1, max_denominator + 1):
        n = round(decimal * d)
        if n == 0:
            continue
        error = abs(decimal - n / d)

        if error < min_error:
            min_error = error
            best_numerator = n
            best_denominator = d

    # Simplify the fraction
    gcd_value = gcd(abs(best_numerator), best_denominator)
    return f"{int(best_numerator / gcd_value)}/{int(best_denominator / gcd_value)}"


def fraction_to_cents(fraction: str) -> float:
    """
    Converts a fraction to cents.
    
    Args:
        fraction: Fraction string (e.g., "3/2")
        
    Returns:
        The equivalent value in cents
        
    Examples:
        >>> fraction_to_cents("3/2")
        701.955
        >>> fraction_to_cents("2/1")
        1200.0
    """
    numerator, denominator = map(int, fraction.split("/"))
    ratio = numerator / denominator
    return 1200.0 * math.log2(ratio)


def cents_to_fraction(cents: float, max_denominator: int = 99) -> str:
    """
    Converts cents to the closest fraction representation.
    
    Args:
        cents: The cent value to convert
        max_denominator: Maximum denominator for the fraction
        
    Returns:
        The closest fraction as a string
        
    Examples:
        >>> cents_to_fraction(702.0)
        "3/2"
        >>> cents_to_fraction(1200.0)
        "2/1"
    """
    ratio = pow(2, cents / 1200.0)
    return _decimal_to_fraction(ratio, max_denominator)


def decimal_to_cents(decimal: float) -> float:
    """
    Converts a decimal ratio to cents.
    
    Args:
        decimal: The decimal ratio
        
    Returns:
        The equivalent value in cents
        
    Examples:
        >>> decimal_to_cents(1.5)
        701.955
        >>> decimal_to_cents(2.0)
        1200.0
    """
    if decimal <= 0:
        raise ValueError("Cannot convert non-positive ratio to cents")
    return 1200.0 * math.log2(decimal)


def cents_to_decimal(cents: float) -> float:
    """
    Converts cents to a decimal ratio.
    
    Args:
        cents: The cent value
        
    Returns:
        The equivalent decimal ratio
        
    Examples:
        >>> cents_to_decimal(702.0)
        1.4983070768766815
        >>> cents_to_decimal(1200.0)
        2.0
    """
    return pow(2, cents / 1200.0)


def midi_to_frequency(midi_note: int, a4_frequency: float = 440.0) -> float:
    """
    Converts a MIDI note number to its frequency in Hz.
    
    Args:
        midi_note: MIDI note number (0-127)
        a4_frequency: Frequency of A4 in Hz (default: 440.0)
        
    Returns:
        The frequency in Hz
        
    Examples:
        >>> midi_to_frequency(69)  # A4
        440.0
        >>> midi_to_frequency(60)  # C4
        261.626
    """
    return a4_frequency * pow(2, (midi_note - 69) / 12.0)


def frequency_to_midi(frequency: float, a4_frequency: float = 440.0) -> int:
    """
    Converts a frequency in Hz to the closest MIDI note number.
    
    Args:
        frequency: Frequency in Hz
        a4_frequency: Frequency of A4 in Hz (default: 440.0)
        
    Returns:
        The closest MIDI note number
        
    Examples:
        >>> frequency_to_midi(440.0)
        69  # A4
        >>> frequency_to_midi(261.626)
        60  # C4
    """
    if frequency <= 0:
        raise ValueError("Frequency must be positive")
    
    midi_float = 69 + 12 * math.log2(frequency / a4_frequency)
    return int(round(midi_float))
